clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java

```diff
@@ -125,13 +128,163 @@
+        assertEquals(invalidRecordNumber - 1, records.records(tp0).size());
+        long lastOffset = records.records(tp0).get(records.records(tp0).size() - 1).offset();
+        assertEquals(invalidRecordNumber - 2, lastOffset);
+
+        consumer.close(0, TimeUnit.MILLISECONDS);
+    }
+
+    @Test
+    public void testSecondPollWithDeserializationErrorThrowsRecordDeserializationException() {
+        // raise an exception from somewhere in the middle of the fetch response
+        // so that we can verify that our position does not advance after raising
+        int invalidRecordNumber = 4;
+        int invalidRecordOffset = 3;
+        ByteArrayDeserializer deserializer = mockErrorDeserializer(invalidRecordNumber);
+
+        KafkaConsumer<String, String> consumer = setUpConsumerAndCreateRecordsToPoll(tp0, 5, deserializer).consumer;
+        ConsumerRecords<String, String> records = consumer.poll(Duration.ZERO);
+
+        assertEquals(invalidRecordNumber - 1, records.count());
+        assertEquals(new HashSet<>(Collections.singletonList(tp0)), records.partitions());
+        assertEquals(invalidRecordNumber - 1, records.records(tp0).size());
+        long lastOffset = records.records(tp0).get(records.records(tp0).size() - 1).offset();
+        assertEquals(invalidRecordNumber - 2, lastOffset);
+
+        try {
+            consumer.poll(Duration.ZERO);
+            fail("Second poll should raise " + RecordDeserializationException.class.getName());
+        } catch (RecordDeserializationException rde) {
+            assertEquals(invalidRecordOffset, rde.offset());
+            assertEquals(tp0, rde.partition());
+        }
+
+        consumer.close(0, TimeUnit.MILLISECONDS);
+    }
+
+
+    @Test
+    public void testPollCorruptedRecordThrowsInoperativeRecordException() {
+        int numRecords = 10;  // total number is 20
+        int corruptRecordOffset = 19;
+
+        ConsumerContainer container = setUpConsumerAndCreateRecordsToPoll(tp0, numRecords);
+        // create corrupt records
+        container.client.prepareResponseFrom(fetchResponse(tp0, 10, numRecords, false, true), container.node);
+        KafkaConsumer<String, String> consumer = container.consumer;
+        // 1-19 clean; 20 corrupt
+        consumer.poll(Duration.ZERO); // eat up first 19 clean records
+
+        try {
+            consumer.poll(Duration.ZERO);
+            fail("Poll should raise " + InoperativeRecordException.class.getName());
+        } catch (InoperativeRecordException ire) {
+            assertEquals(tp0, ire.partition());
+            assertEquals(corruptRecordOffset, ire.offset());
+        }
+
+        consumer.close(0, TimeUnit.MILLISECONDS);
+    }
+
+    /*
+        Create a mock deserializer which throws a SerializationException on the Nth record's value deserialization
+     */
+    private ByteArrayDeserializer mockErrorDeserializer(int recordNumber) {
```

Comment:
Not sure if its better to have these helper methods at the very bottom of the test class or right below where they're used